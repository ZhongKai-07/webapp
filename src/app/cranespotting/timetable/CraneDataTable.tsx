import { Feature, FeatureCollection, Point } from 'geojson';

import ToggleSlider from '@/lib/ToggleSlider';

interface CraneTableRowProps {
  crane: Feature<Point>;
  isSelected: boolean;
  // onSelect: (timestamp: number) => void;
  onSelect: () => void;
  count: number;
}

interface CraneDataTableProps {
  cranesFilteredByDay?: FeatureCollection<Point>;
  allCranes?: FeatureCollection<Point>;
  // selectedTimestamp: number;
  selectedTimestamps: number[];
  // onSelect: (timestamp: number) => void;
  onSelect: (timestamp: number[]) => void;
  testId?: string;
}

const CraneTableRow = ({
  crane,
  isSelected,
  onSelect,
  count,
}: CraneTableRowProps) => (
  <tr className='border-b-2 border-slate-750 rounded'>
    <td className='whitespace-nowrap px-6 py-4 text-sm font-medium text-gray-900'>
      <ToggleSlider checked={isSelected} onClick={onSelect} />
    </td>
    <td className='whitespace-nowrap px-6 py-4 text-lg font-medium text-gray-900'>
      {new Date(crane.properties?.timestamp * 1000).toLocaleTimeString()}
    </td>
    <td className='whitespace-nowrap px-6 py-4 text-lg text-gray-900'>
      {count} crane{count > 1 ? 's' : ''}
    </td>
  </tr>
);

// <input
//        type="radio"
//        name="selectedTime"
//        checked={isSelected}
//        onChange={onSelect}
//      />

export const CraneDataTable: React.FC<CraneDataTableProps> = ({
  cranesFilteredByDay,
  // selectedTimestamp,
  selectedTimestamps,
  onSelect,
  allCranes,
  testId,
}) => {
  return (
    <table style={{ width: '100%' }} data-testid={testId}>
      <thead className='bg-gray-100'>
        <tr>
          <th className='px-6 py-3 text-left text-xs font-medium uppercase tracking-wider text-gray-500'>
            Select
          </th>
          <th className='px-6 py-3 text-left text-xs font-medium uppercase tracking-wider text-gray-500'>
            Time
          </th>
          <th className='px-6 py-3 text-left text-xs font-medium uppercase tracking-wider text-gray-500'>
            Cranes detected
          </th>
        </tr>
      </thead>
      <tbody className='bg-white'>
        {cranesFilteredByDay?.features?.map((crane, index) => (
          <CraneTableRow
            key={index}
            crane={crane}
            isSelected={selectedTimestamps.includes(
              crane.properties?.timestamp
            )}
            // isSelected={crane.properties?.timestamp === selectedTimestamp}
            onSelect={() =>
              onSelect(
                selectedTimestamps.includes(crane.properties?.timestamp)
                  ? selectedTimestamps.filter(
                      (timestamp) => timestamp != crane.properties?.timestamp
                    )
                  : [...selectedTimestamps, crane.properties?.timestamp]
              )
            }
            // onSelect
            // }
            count={
              allCranes?.features.filter(
                (feature) =>
                  feature?.properties?.timestamp === crane.properties?.timestamp
              ).length || 0
            }
          />
        ))}
      </tbody>
    </table>
  );
};
