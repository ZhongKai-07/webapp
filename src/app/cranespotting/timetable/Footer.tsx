export const Footer: React.FC<{ lastRefreshTime: string; testId?: string }> = ({
  lastRefreshTime,
  testId,
}) => (
  <div
    className='bg-white p-2 text-center text-sm text-gray-600'
    data-testid={testId}
  >
    Last Data refresh: {lastRefreshTime}
  </div>
);
