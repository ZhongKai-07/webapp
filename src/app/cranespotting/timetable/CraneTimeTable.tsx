import { Feature, FeatureCollection, Point } from 'geojson';
import React, { useEffect, useState } from 'react';

import TableLoading from '@/lib/TableLoading';

import { CraneDataTable } from '@/app/cranespotting/timetable/CraneDataTable';
import { DateSelector } from '@/app/cranespotting/timetable/DateSelector';
import { Footer } from '@/app/cranespotting/timetable/Footer';

interface CraneTimeTableProps {
  cranes?: FeatureCollection<Point>;
  craneTimestamps: number[];
  setCraneTimestamps: React.Dispatch<React.SetStateAction<number[]>>;
  testId?: string;
}

export const isOnSameDay = (
  date1Num: number | undefined,
  date2Num: number | undefined
) => {
  const date1 = date1Num ? new Date(date1Num * 1000) : null;
  const date2 = date2Num ? new Date(date2Num * 1000) : null;
  if (!date1 || !date2) return false;
  return (
    date1.getDate() === date2.getDate() &&
    date1.getMonth() === date2.getMonth() &&
    date1.getFullYear() === date2.getFullYear()
  );
};

export const calculateTimestampsByDay = (timestamps: number[]) =>
  timestamps.filter((timestamp, index) => {
    const currentDate = new Date(timestamp * 1000);
    const previousDate =
      index > 0 ? new Date(timestamps[index - 1] * 1000) : null;
    return (
      previousDate === null ||
      currentDate.getDate() !== previousDate.getDate() ||
      currentDate.getMonth() !== previousDate.getMonth() ||
      currentDate.getFullYear() !== previousDate.getFullYear()
    );
  });

const CraneTimeTable: React.FC<CraneTimeTableProps> = ({
  cranes,
  craneTimestamps,
  setCraneTimestamps,
  testId,
}) => {
  const [deduplicatedCranes, setDeduplicatedCranes] =
    useState<FeatureCollection<Point>>();
  const [displayedCranes, setDisplayedCranes] =
    useState<FeatureCollection<Point>>();
  const [craneDates, setCraneDates] = useState<number[]>([]);
  const [craneSelectedDay, setCraneSelectedDay] = useState<number>();

  useEffect(() => {
    if (craneSelectedDay && deduplicatedCranes) {
      setDisplayedCranes({
        type: 'FeatureCollection',
        features: deduplicatedCranes.features.filter((feature) =>
          isOnSameDay(feature.properties?.timestamp, craneSelectedDay)
        ),
      });
    }
  }, [craneSelectedDay, deduplicatedCranes]);

  useEffect(() => {
    const timestamps = Array.from(
      new Set(
        (cranes as FeatureCollection<Point>)?.features
          .map((feature) => feature.properties?.timestamp)
          .filter((el) => !!el)
      )
    ) as number[];
    const deduplicatedCranes: FeatureCollection<Point> = {
      type: 'FeatureCollection',
      features: timestamps
        .map((timestamp) =>
          cranes?.features.find(
            (feature) => feature.properties?.timestamp === timestamp
          )
        )
        .filter(Boolean) as Feature<Point>[],
    };
    setDeduplicatedCranes(deduplicatedCranes);
    setDisplayedCranes(deduplicatedCranes);

    const timestampsByDay = calculateTimestampsByDay(timestamps);
    setCraneDates(timestampsByDay);
    setCraneSelectedDay(timestampsByDay[0]);
  }, [cranes]);

  if (!cranes) return <TableLoading />;
  return (
    <div className='flex flex-col' data-testid={testId}>
      <DateSelector
        dates={craneDates}
        selectedDay={craneSelectedDay}
        onSelect={setCraneSelectedDay}
        testId='date-selector'
      />
      <CraneDataTable
        cranesFilteredByDay={displayedCranes}
        allCranes={cranes}
        selectedTimestamps={craneTimestamps}
        onSelect={setCraneTimestamps}
        testId='crane-data-table'
      />
      <Footer
        lastRefreshTime={new Date(craneTimestamps[0] * 1000).toLocaleString()}
        testId='footer'
      />
      <div className='flex w-full justify-center'>
        <button
          className='rounded-full w-1/3 content-center  px-5 py-1 text-base font-medium text-gray-700 border-2 border-apian-yellow active:bg-apian-yellow hover:bg-apian-yellow-opacity-50'
          onClick={() => {
            setCraneTimestamps([]);
          }}
        >
          Clear Selections - (
          {
            cranes?.features?.filter((feature) =>
              craneTimestamps.includes(feature.properties?.timestamp)
            ).length
          }{' '}
          selected)
        </button>
      </div>
    </div>
  );
};

export default CraneTimeTable;
