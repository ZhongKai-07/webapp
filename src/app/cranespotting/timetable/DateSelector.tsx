import { useState } from 'react';

import { isOnSameDay } from '@/app/cranespotting/timetable/CraneTimeTable';

interface DateButtonProps {
  date: number;
  isSelected: boolean;
  onSelect: (timestamp: number) => void;
}

interface DateSelectorProps {
  dates: number[];
  selectedDay?: number;
  onSelect: (timestamp: number) => void;
  testId?: string;
}
const DateButton = ({ date, isSelected, onSelect }: DateButtonProps) => {
  const [pressed, setPressed] = useState<boolean>(false);
  return (
    <button
      className={`rounded-full px-5 py-1 text-base font-medium text-gray-700 border-2 border-apian-yellow ${
        isSelected
          ? 'bg-apian-yellow'
          : 'bg-white hover:bg-apian-yellow-opacity-50'
      } ${pressed ? 'shadow-inner shadow' : 'shadow'}`}
      key={date}
      onMouseDown={() => setPressed(true)}
      onMouseUp={() => setPressed(false)}
      onClick={() => onSelect(date)}
    >
      {new Date(date * 1000).toLocaleDateString()}
    </button>
  );
};

export const DateSelector: React.FC<DateSelectorProps> = ({
  dates,
  selectedDay,
  onSelect,
  testId,
}) => (
  <div className='flex justify-center space-x-2 p-3' data-testid={testId}>
    {dates.map((date) => (
      <DateButton
        key={date}
        date={date}
        isSelected={isOnSameDay(date, selectedDay)}
        onSelect={onSelect}
      />
    ))}
  </div>
);
