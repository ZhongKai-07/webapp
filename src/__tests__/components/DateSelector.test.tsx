import { cleanup, render, screen } from '@testing-library/react';

import { DateSelector } from '@/app/cranespotting/timetable/DateSelector';

const origDate = global.Date.prototype.toLocaleDateString;
jest
  .spyOn(global.Date.prototype, 'toLocaleDateString')
  .mockImplementation(function () {
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    return origDate.call(this, 'en-GB');
  });

describe('DateSelector', () => {
  const timestamps = [1627833600, 1627920000, 1628006400];
  const selectedDay = timestamps[1];
  afterEach(cleanup);

  it('should render correctly', () => {
    const onSelect = jest.fn();
    render(
      <DateSelector
        dates={timestamps}
        selectedDay={selectedDay}
        onSelect={onSelect}
      />
    );

    expect(screen.getByText('01/08/2021')).toBeInTheDocument();
    expect(screen.getByText('02/08/2021')).toBeInTheDocument();
    expect(screen.getByText('03/08/2021')).toBeInTheDocument();
    expect(screen.getByText('02/08/2021')).toHaveClass(
      'rounded-full px-5 py-1 text-base font-medium text-gray-700 border-2 border-apian-yellow bg-apian-yellow shadow'
    );
  });

  it('should call onSelect when a date button is clicked', () => {
    const onSelect = jest.fn();
    render(
      <DateSelector
        dates={timestamps}
        selectedDay={selectedDay}
        onSelect={onSelect}
      />
    );

    const dateButton = screen.getByText('03/08/2021');
    dateButton.click();

    expect(onSelect).toHaveBeenCalledWith(1628006400);
  });

  it('should match the snapshot', () => {
    const onSelect = jest.fn();
    const { container } = render(
      <DateSelector
        dates={timestamps}
        selectedDay={selectedDay}
        onSelect={onSelect}
      />
    );

    expect(container).toMatchSnapshot();
  });
});
