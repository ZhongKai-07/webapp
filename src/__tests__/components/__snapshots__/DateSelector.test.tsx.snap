// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateSelector should match the snapshot 1`] = `
<div>
  <div
    class="flex justify-center space-x-2 p-3"
  >
    <button
      class="rounded-full px-5 py-1 text-base font-medium text-gray-700 border-2 border-apian-yellow bg-white hover:bg-apian-yellow-opacity-50 shadow"
    >
      01/08/2021
    </button>
    <button
      class="rounded-full px-5 py-1 text-base font-medium text-gray-700 border-2 border-apian-yellow bg-apian-yellow shadow"
    >
      02/08/2021
    </button>
    <button
      class="rounded-full px-5 py-1 text-base font-medium text-gray-700 border-2 border-apian-yellow bg-white hover:bg-apian-yellow-opacity-50 shadow"
    >
      03/08/2021
    </button>
  </div>
</div>
`;
